// app/honey/page/web/components/Apply/index.js
import React, {Component} from 'react'
import {Grid, FormGroup, Typography} from '@material-ui/core';
import compose from 'recompose/compose'
import withStateHandlers from 'recompose/withStateHandlers'
import defaultProps         from 'recompose/defaultProps'
import FormControl from '@material-ui/core/FormControl';

import {
  ButtonPrimary,
  TypographyItalic,
  Input,
} from '../../customUI.js';

import {MinusIcon, PlusIcon} from '../../icons';
import {rem} from 'polished'
import styled from 'styled-components'

function Detailes({ children }) {
  return (
    <SubTitle component="div" style={{ padding: 8 * 3 }} >
      {children}
    </SubTitle>
  );
}

const SubTitle = styled(({ ...other }) => (
  <Typography {...other} />
))`
&& {
  color: #e1b263;
  font-size: ${rem(24)};
  text-align: center;
  line-height: 1.4;
  font-weight: normal;
  font-style: italic;
  margin: 0;
}
`;
const enhance = compose(
  defaultProps({
    Input: styled(Input)`
      && {
        input {
          font-weight: 600;
        }
      }
    `,
    MinusIcon: styled(MinusIcon)`
      && {
      }
    `,
    PlusIcon: styled(PlusIcon)`
      && {

      }
    `,
    FormGroup: styled(FormGroup)`
      && {
      }
  `
  }),
  withStateHandlers(
    {
      count: 2
    },
    {
      increment: ({ count }) => () => ({ count: count + 1 }),
      decrement: ({ count }) => () => {
        return count > 2 && { count: count - 1 }
      }
    }
  )
)

const Apply = enhance(({
  count,
  direction,
  decrement,
  increment,
  Input,
  MinusIcon,
  PlusIcon
}) => {
  return (
    <Grid container direction={'column'} justify="center" alignItems="center" spacing={0}>
      <Grid xs={10} sm={5} md={4} item>

        <TypographyItalic align="center" size="0.875em" bold="300" color="#000" margin="0 auto 0.9375rem" variant="subtitle1">
          Receive twelve jars of 850 g harvested and filled on site
        </TypographyItalic>

        <Grid container alignItems="center" justify="center" spacing={0}>

          <Grid item xs={2}>
            <MinusIcon onClick={decrement} color="primary"/ >
          </Grid>   

          <Grid item xs={6}>

            <Input
                variant="standard"
                name="value"
                label=""
                disabled
                id="standard-name"
                value={count+ ' beehives'}
                align="center"
                width="100%"
              /> 
          </Grid>

          <Grid item xs={2} style={{textAlign: 'right'}}>
            <PlusIcon onClick={increment} color="primary"/>
          </Grid>   

          <Grid item xs={10}>
            <Detailes dir={direction}>
              {'$'}{35 * count}
            </Detailes>

          </Grid>


          <Grid item xs={10}>
            <ButtonPrimary fullWidth variant="contained" color="primary">Apply</ButtonPrimary>
          </Grid>

        </Grid>


      </Grid>

    </Grid>
  )
})

export default Apply
