// box/2d/point/web/Point2d.js
import * as PIXI            from 'pixi.js'

import React, { Component } from 'react'
import compose              from 'recompose/compose'
import defaultProps         from 'recompose/defaultProps'
import withHandlers         from 'recompose/withHandlers'
import withPropsOnChange    from 'recompose/withPropsOnChange'
import withStateHandlers    from 'recompose/withStateHandlers'
import lifecycle            from 'recompose/lifecycle'


import withLayer            from '../../layer/all/withLayer'
import withStage            from '../../stage/all/withStage'


const enhance = compose(
  withLayer(),
  withStage(),
  defaultProps({
    fillColor: 0xffffff,
    color: 0x0000ff,
    radius: 4,
    lineAlpha: 1,
    lineWidth: 2,
    x: 0,
    y: 0,
  }),
  withStateHandlers(
    () => ({
      instance: null,
    }),
    {
      instanceSet: () => (instance) => ({ instance }),
    }
  ),
  withPropsOnChange(
    ['x', 'y'],
    ({ instance, fromStage, x, y, z }) => {
      if (instance) {
        instance.position.set(x, y)
        fromStage.stage.dirty = true
      }
    }
  ),
  withHandlers({
    update: ({
      color,
      fillColor,
      instance,
      lineAlpha,
      lineWidth,
      radius,
    }) => (instanceForced) => {
      const o = instance || instanceForced
      if (o) {
        o.beginFill(fillColor)
        o.lineStyle(lineWidth, color, lineAlpha)
        o.drawCircle(0, 0, radius)
        o.endFill()
      }
    },
  }),
  withHandlers({
    mount: ({
      fromLayer,
      fromStage,
      instanceSet,
      update,
      x, y,
    }) => () => {
      const { stage } = fromStage
      const instance = new PIXI.Graphics()
      instance.position.set(x, y)
      instance.parentGroup = fromLayer.layer
      stage.addChild(instance)
      instanceSet(instance)
      update(instance)
    },
    unmount: ({
      fromStage,
      instance,
    }) => () => {
      const { stage } = fromStage
      if (stage && instance) {
        stage.removeChild(instance)
      }
    }
  }),
  lifecycle({
    componentDidMount() {
      this.props.mount()
    },
    componentDidUpdate() {
      //this.props.update()
    },
    componentWillUnmount() {
      this.props.unmount()
    },
  }),
)

const Point2d = enhance(() => null)

export default Point2d
