// app/honey/page/web/components/SocialThumbs/index.js
import React, {Component} from 'react'
import {Grid, Typography} from '@material-ui/core'
import compose from 'recompose/compose'
import './main.scss'
import styled from 'styled-components'
import {rem} from 'polished'
import {IconFacebook, IconInstagram, IconTwitter, IconYoutube} from '../../icons'

const enhance = compose()

const Title = styled(({ color, transform, border, margin, fweight, ...other }) => (
  <Typography {...other} />
))`&& {
  color: ${props => props.color};
  border-bottom: ${props => props.border};
  text-transform: ${props => props.transform || 'inherit'};
  border-color: #000;
  margin: ${props => props.margin};
  font-weight: ${props => props.fweight || '700'};
  min-width: auto;
  width: 100%;
  text-align: center;
  font-size: ${rem(35)};
}
`;

const SocialThumbnailLink = styled.a`
  position: relative;
  width: 100%;
  height: 100%;
  text-align: right;
  display: block;
  overflow: hidden;
  border-right: 3px solid white;
`;

const SocialThumbnailImg = styled.img`
  width: 100%;
  height: 100%;
  text-align: right;
  display: block;
  overflow: hidden;
  border-right: 3px solid white;
`;

const SocialIcon = styled.div`
  display: inline-block;
  position: absolute;
  bottom: ${rem(20)};
  right: ${rem(20)};
`;

const iconMap = {
  'instagram': IconInstagram,
  'fb': IconFacebook,
  'youtube': IconYoutube,
  'tw': IconTwitter
}

const Social = enhance(() => {
  return (
    <div>
      <Title transform="uppercase" margin="0 0 2.25rem 0">
        Sozial Buzz
      </Title>

      <Grid container direction={'row'} justify="space-between" alignItems={'baseline'} spacing={0}>
        {Object.keys(iconMap).map((item, index) => {
          let Icon = iconMap[item];
          return (
            <Grid md={3} sm={6} xs={12} key={index} item className={'socialThumbnail'}>
              <SocialThumbnailLink href={`//www.instagram.com/p/Bp9tfPzDdbQ/`}
                                   target="_blank"
                                   rel="noopener noreferrer">
                <SocialThumbnailImg src={'/img/social' + (index + 1) + '.jpg?v=1'} alt=""/>
                <SocialIcon>
                  <Icon style={{width: 'auto', height: '30px', fill: '#fff'}}/>
                </SocialIcon>
              </SocialThumbnailLink>
            </Grid>
          )
        })}
      </Grid>
    </div>
  )
})

export default Social
