// box/2d-editor/editor/web/block/BlockEditor.js
import { Grid }             from '@material-ui/core'

import React, { Component } from 'react'
import compose              from 'recompose/compose'
import withHandlers         from 'recompose/withHandlers'
import withStateHandlers    from 'recompose/withStateHandlers'
import styled               from 'styled-components'

import { Button }           from 'semantic-ui-react'
import { Icon }             from 'semantic-ui-react'


import controlTypeMap       from 'box/3d/control/api/controlTypeMap'
import withSpace            from 'evoke-me/space/all/withSpace'

import statusMap            from '../../api/statusMap'
import spaceEditor          from '../../space/spaceEditor'
import ControlType          from '../ControlType'
import Editor               from '../Editor'
import Status               from '../Status'


const Block = styled.div`
  width: 100%;
  height: 100%;
`

const StatusStyled = styled(Status)`
  position: absolute;
  top: 10px;
  right: 10px;

  user-select: none;
`

const ControlTypeStyled = styled(ControlType)`
  position: absolute;
  bottom: 10px;
  right: 10px;

  user-select: none;
`

const GridStyled = styled(Grid)`
  padding-top: 5px;
`

const enhance = compose(
  withSpace(spaceEditor),
  withStateHandlers(() => ({
    textureIndex: 0,
  }), {
    textureIndexSet: () => (textureIndex) => ({ textureIndex }),
  }),
)

const BlockEditor = enhance(({
  fromEditor,
  textureIndex, textureIndexSet,
}) => {
  const { status } = fromEditor
  const { controlType } = fromEditor

  return (
    <Block>
      <Editor
        controlType={controlType}
        status={status}
        textureIndex={textureIndex}
      >
        <StatusStyled status={status} />
        <ControlTypeStyled controlType={controlType} />
      </Editor>
      <GridStyled container justify='center' spacing={8}>
        <Grid item>
          <Button
            animated='fade'
            fluid
            onClick={() => fromEditor.statusSet(statusMap.AREA_CREATE)}
          >
            <Button.Content hidden>AREA</Button.Content>
            <Button.Content visible>
              <Icon name='plus' />
            </Button.Content>          
          </Button>
        </Grid>
        <Grid item>
          <Button
            animated='fade'          
            onClick={() => fromEditor.statusSet(statusMap.DEFAULT)}
          >
            <Button.Content hidden>Finish</Button.Content>
            <Button.Content visible>
              <Icon name='times rectangle' />
            </Button.Content>          
          </Button>
        </Grid>        
        <Grid item>
          <Button
            animated='fade'
            onClick={() => fromEditor.controlTypeSet(controlTypeMap.ORBITAL)}
          >
            <Button.Content hidden>Control</Button.Content>
            <Button.Content visible>
              <Icon name='circle outline' />
            </Button.Content>
          </Button>
        </Grid>        
      </GridStyled>
      <GridStyled container justify='center' spacing={8}>
        <Grid item>
          <Button
            animated='fade'
            onClick={() => textureIndexSet(0)}
          >
            <Button.Content hidden>0</Button.Content>
            <Button.Content visible>
              <Icon name='file image' />
            </Button.Content>
          </Button>
        </Grid>
        <Grid item>
          <Button
            animated='fade'
            onClick={() => textureIndexSet(1)}
          >
            <Button.Content hidden>1</Button.Content>
            <Button.Content visible>
              <Icon name='file image' />
            </Button.Content>
          </Button>
        </Grid>
        <Grid item>
          <Button
            animated='fade'
            onClick={() => textureIndexSet(2)}
          >
            <Button.Content hidden>2</Button.Content>
            <Button.Content visible>
              <Icon name='file image' />
            </Button.Content>
          </Button>
        </Grid>
      </GridStyled>
    </Block>
  )
})

export default BlockEditor
