// box/3d/scene/web/block/BlockScene.js
import Button               from '@material-ui/core/Button'
import Grid                 from '@material-ui/core/Grid'

import React, { Component } from 'react'
import compose              from 'recompose/compose'
import withStateHandlers    from 'recompose/withStateHandlers'
import styled               from 'styled-components'


import Image2d              from 'box/2d/image/web/Image2d'
import Layer2d              from 'box/2d/layer/web/Layer2d'
import Line2d               from 'box/2d/line/web/Line2d'
import Point2d              from 'box/2d/point/web/Point2d'
import Stage                from 'box/2d/stage/web/Stage'

import Camera3d             from 'box/3d/camera/web/Camera3d'
import Grid3d               from 'box/3d/grid/web/Grid3d'
import LightAmbient3d       from 'box/3d/lightAmbient/web/LightAmbient3d'
import LightPoint3d         from 'box/3d/lightPoint/web/LightPoint3d'
import Object3d             from 'box/3d/object3d/web/Object3d'
import Plane3d              from 'box/3d/plane/web/Plane3d'
import Scene                from 'box/3d/scene/web/Scene'
import Sphere3d             from 'box/3d/sphere/web/Sphere3d'
import Video3d              from 'box/3d/video/web/Video3d'



const Block = styled.div`
  width: 100%;
  height: 100%;
`

const enhance = compose(
  withStateHandlers(
    () => ({
      planeInstance: null,
      canvasTexture1: null,
      width: 2000,
      height: 2000,
    }),
    {
      set: () => (x, y) => ({ x, y }),
      left: ({ x }) => () => ({ x: x - 5 }),
      right: ({ x }) => () => ({ x: x + 5 }),
      canvasTexture1Set: () => (canvasTexture1) => ({ canvasTexture1 }),
      planeInstanceSet: () => (planeInstance) => ({ planeInstance }),
    }
  )
)

const BlockScene = enhance(({
  canvasTexture1, canvasTexture1Set,
  children,
  left,
  planeInstance, planeInstanceSet,
  right,
  set,
  x, y, z,
  width, height,
}) => {
  return (
    <Block>
      <Scene>
        <Camera3d
          control='plane'
          intersectCheckList={[planeInstance]}
          plane={planeInstance}
          onIntersect={set}
        />
        <LightAmbient3d />
        <LightPoint3d />
        <Plane3d
          canvas={canvasTexture1}
          onInstanceChange={planeInstanceSet}  
        />
        <Object3d
          mtl='https://cdn.evokeme.io/static/3d/1.mtl'
          obj='https://cdn.evokeme.io/static/3d/1.obj'
        />
      </Scene>
      <Stage onCanvasChange={canvasTexture1Set} isShow={false}>
        <Layer2d name='background' index={0}>
          <Image2d
            imageLink='https://i.imgur.com/OynwAM5.jpg'
            isFit
          />
        </Layer2d>
        <Layer2d name='service' index={10}>
          <Point2d x={x * width} y={y * height} />
        </Layer2d>
      </Stage>
    </Block>
  )
})

      // <Grid container justify='center'>
      //   <Grid item>
      //     <Button onClick={left}>LEFT</Button>
      //   </Grid>
      //   <Grid item>
      //     <Button onClick={right}>RIGHT</Button>
      //   </Grid>
      // </Grid>

export default BlockScene
