// evoke-me/space/api/spaceDefaults.js
import uuid from 'uuid/v4'


export default {
  state: {
    data: {},
    orderMap: {
      default: [],
    },
  },
  updaters: {
    create: (getState, setState) => function* (doc, options = {}) {
      const { orderName } = options
      const { data, orderMap } = yield getState()

      const id = doc.id || uuid()

      const orderMapNext = data[id]
        ? { default: orderMap.default }
        : { default: [...orderMap.default, id] }

      if (orderName) {
        orderMapNext[orderName] = [...(orderMap[orderName] || []), id]
      }

      yield setState({
        data: { ...data, [id]: doc },
        orderMap: { ...orderMap, ...orderMapNext },
      })
    },
    update: (getState, setState) => function* (doc) {
      const { data, orderMap } = yield getState()

      if (!data[doc.id]) return null

      yield setState({
        data: {
          ...data,
          [doc.id]: {
            ...data[doc.id],
            ...doc,
          }
        }
      })
    },
    remove: (getState, setState) => function* ({ id }, options = {}) {
      const { orderName = `default` } = options
      const { data: prevData, orderMap } = yield getState()
      const { [id]: docRemoved, ...data } = prevData

      yield setState({
        data,
        orderMap: {
          ...orderMap,
          [orderName]: orderMap[orderName].filter((_id) => _id != id)
        }
      })
    },
    orderClear: (getState, setState) => function* (name) {
      const { orderMap: orderMapPrev } = yield getState()
      const { [name]: orderSelected, ...orderMap } = orderMapPrev

      yield setState({ orderMap })
    },
    orderMove: (getState, setState) => function* (id, from, to) {
      const { orderMap } = yield getState()

      const fromOrder = orderMap[from].filter((_id) => _id != id)
      const toOrder = [...(orderMap[to] || []), id]

      yield setState({
        orderMap: {
          ...orderMap,
          [from]: fromOrder,
          [to]: toOrder,
        }
      })
    },
  }
}
