// app/polyvore/collage/space/spaceCollage.js
import config               from '../../config/api/config'
import uuid                 from 'uuid/v4'

export default [
  config.namespace,
  `collage`,
  [],
  () => ({
    isSeeded: false,
    data: {},
    orderMap: {
      default: [],
    }
  }),
  {
    create: (getState, setState) => function* (doc, options = {}) {
      const { orderName = `default` } = options
      const { data, orderMap } = yield getState()

      const id = doc.id || uuid()

      const orderNext = data[id]
        ? orderMap[orderName] || []
        : [...orderMap[orderName], id]

      yield setState({
        data: {
          ...data,
          [id]: {
            ...(data[id] || {}),
            ...doc,
          },
        },
        orderMap: {
          ...orderMap,
          [orderName]: orderNext,
        }
      })
    },
    update: (getState, setState, { create }) => function* (doc, options = {}) {
      yield create(doc, options)
    },
    remove: (getState, setState) => function* ({ id }, options = {}) {
      const { orderName = `default` } = options
      const { data: prevData, orderMap } = yield getState()
      const { [id]: docRemoved, ...data } = prevData

      yield setState({
        data,
        orderMap: {
          ...orderMap,
          [orderName]: orderMap[orderName].filter((_id) => _id !== id)
        }
      })
    },
    orderMove: (getState, setState) => function* (id, to, from = `default`) {
      const { orderMap } = yield getState()

      const fromOrder = orderMap.default.filter((_id) => _id !== id)
      const toOrder = [
        ...(orderMap[to] || []),
        id,
      ]

      yield setState({
        orderMap: {
          ...orderMap,
          [from]: fromOrder,
          [to]: toOrder,
        }
      })
    },
    seed: (getState, setState, { create }) => function* () {
      const { isSeeded } = yield getState()

      if (!isSeeded) {
        yield create({
          id: `collage1`,
          name: `collage1`,
          imageLink: `https://static1.squarespace.com/static/5335412de4b0ce89f434f462/5a198edb53450a9c542547db/5a198f0ce4966b1342d3460b/1513715870334/8.jpg`,
        })

        yield create({
          id: `collage2`,
          name: `collage2`,
          imageLink: `http://ic.pics.livejournal.com/aynart/19984708/1231936/1231936_original.jpg`,
        })

        yield create({
          id: `collage3`,
          name: `collage3`,
          imageLink: `http://cs4123.userapi.com/u2102074/35996800/y_1555595d.jpg`,
        })

        yield create({
          id: `collage4`,
          name: `collage4`,
          imageLink: `http://2.bp.blogspot.com/-0x9xwNNmhJE/U9pHhR2JNEI/AAAAAAAABOI/US-yHkk57Tg/s1600/%D0%BA%D0%BE%D0%BB%D0%BB%D0%B0%D0%B6+%D0%B1%D0%BB%D0%BE%D0%B3.jpg`,
        })

        yield setState({ isSeeded: true })
      }
    },    
  }
]
