// app/honey/page/web/components/Poster/index.js
import React, {Component} from 'react'
import compose from 'recompose/compose'
import defaultProps         from 'recompose/defaultProps'
import {Grid} from '@material-ui/core';
import {rem} from 'polished'
import styled from 'styled-components'
import {mediaGrid} from '../../customUI'

const enhance = compose(
  defaultProps({
    PosterBlock: styled(Grid)`
      && {
        background-position: center center;
        background-repeat: no-repeat
        background-color: #efefee;
        background-size: 150% auto;
        text-align: center;
        height: ${rem(600)};

        ${mediaGrid(`xl`)}{//lg d
          background-size: 150% auto;
          height: ${rem(600)};
        }
        ${mediaGrid(`lg`)}{//sm d
          background-size: 110% auto;
          height: ${rem(600)};
        }
        ${mediaGrid(`md`)}{//tablet
          background-size: 100% auto;
          height: ${rem(500)};
        }
        ${mediaGrid(`sm`)}{//phone
          background-size: auto 100%;
          height: ${rem(400)};
        }
      }
    `,

    linkItalic: styled.div`
      margin: ${rem(20)} 0 !important;
      a {
        color: #fff;
        text-decoration: underline;
      }
    `,
  }),
)


const Poster = enhance((props) => {
  const {
    children, url, cname, bgcolor, banner, PosterBlock, 
  } = props
  const style = {
    backgroundImage: 'url(' + url + ')',
    backgroundColor: bgcolor || '#efefee'
  };

    return (
      <PosterBlock container direction={'row'} justify="center" alignItems={'center'} spacing={0} className={`${cname} ${banner} `} style={style}  >
        {children}
      </PosterBlock>
    )
  })



export default Poster
