// app/honey/page/web/components/Subscribe/index.js
import React, { Component } from 'react'

import compose              from 'recompose/compose'
import defaultProps         from 'recompose/defaultProps'
import withHandlers         from 'recompose/withHandlers'
import withStateHandlers    from 'recompose/withStateHandlers'

import { Grid, AppBar, Tabs, Typography, Tab, Button} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import {rem} from 'polished'
import {mediaGrid, ButtonPrimary} from '../../customUI'

import SwipeableViews from 'react-swipeable-views';
import styled from 'styled-components';

const ButtonPrimaryCustom = styled(ButtonPrimary)`
  && {
    margin: auto;
    width: 100%;
    ${mediaGrid(`xl`)}{//lg d
      width: 100%;
    }
    ${mediaGrid(`lg`)}{//sm d
      width: 100%;
    }
    ${mediaGrid(`md`)}{//tablet
      width: 100%;
    }
    ${mediaGrid(`sm`)}{//phone
      width: 100%;
    }

   }
`;

const SubTitle = styled(Typography)`
&& {
  white-space: nowrap;
  color: #000;
  font-size: ${rem(14)};
  text-align: center;
  line-height: 1.4;
  font-weight: 300;
  font-style: italic;
  margin: ${rem(40)} -${rem(50)} ${rem(15)};
  margin: 2.5rem auto 0.9375rem;
  padding: 0;

  a {
    text-decoration: underline;
    color: #000;
    &:hover {
      text-decoration: none;
    }
  }

}
`;

const TabContainerSubTitle = styled(Typography)`
&& {
  color: #e1b263;
  font-size: ${rem(24)};
  text-align: center;
  line-height: 1.4;
  font-weight: normal;
  font-style: italic;
}
`;


const AppBarCustom = styled(AppBar)`
&& {
  width: 70%;
  ${mediaGrid(`xl`)}{//lg d
    width: 70%;
  }
  ${mediaGrid(`lg`)}{//sm d
    width: 70%;
  }
  ${mediaGrid(`md`)}{//tablet
    width: 100%;
  }
  ${mediaGrid(`sm`)}{//phone
    width: 70%;
  }
  margin: 0 auto;
}
`;

const TabStyled = styled(Tab)`
&& {
  font-weight: 600;
  font-size: ${rem(18)};
  text-transform: lowercase;
  opacity: inherit;
  min-width: inherit;
  width: inherit;
  border-top: 1px solid rgba(0,0,0,0.2);
  border-bottom: 1px solid rgba(0,0,0,0.2);
  border-right: 1px solid rgba(0,0,0,0.2);
  height: ${rem(48)};
  &:first-child {
    border-left: 1px solid rgba(0,0,0,0.2);
  }
  &:last-child {
    border-right: 1px solid rgba(0,0,0,0.2);
  }
}
`;


const TabContainer = ({ children, dir }) => {
  return (
    <TabContainerSubTitle component="div" dir={dir} style={{ padding: 8 * 3 }}>
      {children}
    </TabContainerSubTitle>
  );
}


const styles = theme => ({
  selected: {
    backgroundColor: '#000000',
    color: '#ffffff',
  },
});


const enhance = compose(
  defaultProps({
  }),
  withStateHandlers(
    () => ({
      value: 0,
      direction: 'rtl'
    }),
    {
      handleChange: () => (event, value) => ({ value }),
      handleChangeIndex: () => (index) => ({value: index})
    }
  ),
  withHandlers(() => {

  }),
)

const Subscribe = enhance(({
  theme, value, direction, handleChange, handleChangeIndex
}) => {
  return (
    <Grid container direction={'column'} justify="flex-start" alignItems="center" spacing={0}>

      <Grid xs={10} sm={5} item >

          <Grid container direction={'column'} justify="flex-start" alignItems="stretch" spacing={0}>
            <Grid item >
                <SubTitle variant="subtitle1">
                  Receive three jars every three months from your selected honey trail
                </SubTitle>

                <AppBarCustom position="static" color="default">
                  <Tabs
                    value={value}
                    onChange={handleChange}
                    fullWidth
                  >
                    <TabStyled  fullWidth label="230 g" />
                    <TabStyled fullWidth label="450 g" />
                    <TabStyled fullWidth label="850 g" />
                  </Tabs>
                </AppBarCustom>


                <SwipeableViews
                  axis={theme.direction === 'rtl' ? 'x-reverse' : 'x'}
                  index={value}
                  onChangeIndex={handleChangeIndex}
                >

                  <TabContainer dir={theme.direction}>
                    {'$ 299 per year'}
                  </TabContainer>

                  <TabContainer dir={theme.direction}>
                    {'$ 499 per year'}
                  </TabContainer>

                  <TabContainer dir={theme.direction}>
                    {'$ 999 per year'}
                  </TabContainer>

                </SwipeableViews>
            </Grid>
            <Grid item >
              <ButtonPrimaryCustom fullWidth variant="contained" color="primary">Subscribe</ButtonPrimaryCustom>
            </Grid>
          </Grid>

      </Grid>

    </Grid>
  )
})

export default withStyles(styles, { withTheme: true })(Subscribe);
