// app/honey/page/web/components/ProductSampleList/index.js
import React, {Component} from 'react'
import {Grid, Typography, FormControl} from '@material-ui/core';
import compose from 'recompose/compose'
import styled from 'styled-components'
import {rem} from 'polished'
import {mediaGrid, ButtonBuy} from '../../customUI'

const ReadMore = styled(({ size, align, margin, bold, color, ...other }) => (
  <Typography {...other} />
))`
&& {
  text-align: center;
  color: ${'#000'};
  font-size: ${props => props.size};
  line-height: 1.4;
  font-weight: ${props => props.bold || 'normal'};
  margin-bottom: ${rem(18)};
  margin: ${props => props.margin || '0'};
  padding: ${props => props.padding || '0'};

  a {
    text-decoration: underline;
    color: ${'#000'};
    font-weight: ${props => props.bold || 'normal'};
    &:hover {
      text-decoration: none;
    }
  }

}
`;

const ButtonBaseImg = styled(FormControl)`
  && {
    text-align: center;
    // margin-right: ${rem(40)};
    width: ${rem(100)};
    justify: center;

    img {
      width: ${rem(100)};
      height: auto;
    }

    ${mediaGrid(`md`)}{//tablet
      text-align: left;
        img {
          width: ${rem(100)};
          height: auto;
        }
      }
    ${mediaGrid(`sm`)}{//phone
      text-align: center;
      img {
        width: ${rem(100)};
        height: auto;
      }
    }
  }
`;


const ProductParam = styled(Typography)`
  && {
    font-size:  ${rem(16)};
    line-height:  ${rem(23)};
    font-style: normal;
    font-stretch: normal;
    line-height: normal;
    letter-spacing: normal; 
    margin-bottom: ${rem(5)};
  }
`;


const Title = styled(Typography)`
  && {
    font-weight: 700;
    font-size: ${rem(17)};
    line-height: normal;
    border-bottom: 1px solid;
    padding-bottom: ${rem(7)};
    margin-bottom: ${rem(13)};
    color: rgba(0, 0, 0, 0.87);
    border-color: rgba(0, 0, 0, 0.87);
    text-transform: uppercase;
  }
`

const SubTitle = styled(Typography)`
  && {
    font-size: ${rem(14)};
    text-align: center;
    width: 160px;
    max-height: 40px;

  }
`



const enhance = compose()

const ProductSampleList = enhance(({}) => {
  return (

  
    <Grid container direction={'row'} justify={'center'} alignItems={'center'} spacing={0}>
      <Grid container
      direction="row"
      justify="center"
      alignItems="center">
      <Title>Loquat and wild flower </Title>
    

      </Grid>

      <Grid item >

          <FormControl className={'details'}>
            <Grid container direction={'column'} alignItems={'stretch'} justify={'space-around'} spacing={0}>
                <Grid item xs>
                 
                  <SubTitle variant="caption">
                    An aromatic winter honey from the Alman Plateau
                  </SubTitle>
                  <ReadMore bold="600" size="10px" margin="0.9375rem auto 0.875rem" variant="button" gutterBottom>
                  <a href="/">Read more</a>
                </ReadMore>  
                  <Grid container  
      direction="row"
      justify="center"
      alignItems="center">
                  <ButtonBaseImg>
          <img  alt={'productTitle'} src="/img/product1.jpg"/>
        </ButtonBaseImg>
        </Grid> 
                  {/* <ProductParam paragraph={true}>
                    <b>Sweetness:</b> 3/5
                  </ProductParam>
                  <ProductParam paragraph={true}>
                    <b>Taste:</b> zest, fruit, nut and spice notes
                  </ProductParam>
                  <ProductParam paragraph={true}>
                    <b>Food pairing:</b> breads, dairy products and desserts
                  </ProductParam>
                  <ProductParam paragraph={true}>
                    <b>Texture:</b> small to fine crystals
                  </ProductParam> */}

                </Grid>
                <Grid item xs>
                  <br/>
                  <ButtonBuy  fullWidth variant="contained"  align='center'  color="primary">Buy Now</ButtonBuy>
                </Grid>

                
            </Grid>

          </FormControl>

      </Grid>
    </Grid>


    )
})

export default ProductSampleList
