// app/honey/page/web/components/PressQuotes/index.js
import React, {Component} from 'react'
import {Grid, ButtonBase, Typography} from '@material-ui/core'
import compose from 'recompose/compose'
import styled from 'styled-components'
import  {rem} from 'polished'

const TypographyItalic = styled(({ size, align, margin, bold, color, ...other }) => (
  <Typography {...other} />
))`
&& {
  color: ${props => props.color || '#fff'};
  font-size: ${props => props.size};
  text-align: ${props => props.align || 'left'};
  line-height: 1.4;
  font-weight: ${props => props.bold || 'normal'};
  font-style: italic;
  margin: ${props => props.margin || '0'};
  padding: ${props => props.padding || '0'};

  a {
    text-decoration: underline;
    color: ${props => props.color || '#fff'};
    font-weight: ${props => props.bold || 'normal'};
    &:hover {
      text-decoration: none;
    }
  }

}
`;

const PressQuotesBlock = styled.div
`&& {
  text-align: center;
`; 

const PressQuotesContainer = styled.div
`&& {
  background: rgba(0,0,0,0.05);
  padding: ${rem(69)} 1.875rem;
`; 


const Press = styled(ButtonBase)`
&& {
  width: 100%;
  height: 130px;
}  
`;

const Title = styled(({ color, transform, border, margin, fweight, ...other }) => (
  <Typography {...other} />
))`&& {
  color: ${props => props.color};
  border-bottom: ${props => props.border};
  text-transform: ${props => props.transform || 'inherit'};
  border-color: ${props => props.color};
  margin: ${props => props.margin};
  font-weight: ${props => props.fweight || '700'};
  min-width: auto;
  width: 100%;
  text-align: center;
  font-size: ${rem(35)};

}
`;

const enhance = compose()

const PressQuotes = enhance(({}) => {
  return (
    <PressQuotesBlock>
      <Title color="black" transform="uppercase" margin="0 0 2.25rem 0">
        Press
      </Title>

      <PressQuotesContainer>
        <Grid item xs={12} sm={10} style={{ 'margin': 'auto' }}>

          <Grid container direction={'row'} justify="space-between" alignItems={'center'} spacing={40}>
            {(Array.apply(null, Array(3))).map((value, index) => (
              <Grid key={index} item sm={4}>

                <Grid container direction={'column'} justify="center" alignItems={'center'} spacing={40}>
                  <Grid item xs>
                    <Press>
                      <img alt={'Press Quote' + index + 1} src={'/img/press' + (index + 1) + '.png'}/>
                    </Press>
                  </Grid>
                  <Grid item xs>
                    <TypographyItalic color="#000" bold="700" align="center" size="1.125rem" variant="subtitle1" gutterBottom>
                      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas fringilla massa vitae nisl
                      scelerisque lobortis.
                    </TypographyItalic>
                  </Grid>
                </Grid>

              </Grid>
            ))}
          </Grid>
        </Grid>
      </PressQuotesContainer>
    </PressQuotesBlock>
  )
})

export default PressQuotes
