// box/2d/line/web/Line2d.js
import * as PIXI            from 'pixi.js'

import React, { Component } from 'react'
import compose              from 'recompose/compose'
import defaultProps         from 'recompose/defaultProps'
import withHandlers         from 'recompose/withHandlers'
import withPropsOnChange    from 'recompose/withPropsOnChange'
import withStateHandlers    from 'recompose/withStateHandlers'
import lifecycle            from 'recompose/lifecycle'


import withLayer            from '../../layer/all/withLayer'
import withStage            from '../../stage/all/withStage'


const enhance = compose(
  withLayer(),
  withStage(),
  defaultProps({
    color: 0x0000ff,
    path: [
      [0, 0],
      [100, 100],
    ],
    x: 0,
    y: 0,
  }),
  withStateHandlers(
    () => ({
      instance: null,
    }),
    {
      instanceSet: () => (instance) => ({ instance }),
    }
  ),
  withPropsOnChange(
    ['x', 'y'],
    ({ instance, x, y, }) => {
      if (instance) {
        instance.position.set(x, y)
      }
    }
  ),
  withHandlers({
    update: ({
      color,
      instance,
      path,
    }) => (o = instance) => {
      if (o) {
        o.clear()
        o.lineStyle(2, 0xff0000, 1)
        for (let [index, [x, y]] of path.entries()) {
          if (index) {
            o.lineTo(x, y)
          } else {
            o.moveTo(x, y)
          }
        }
      }
    },
  }),
  withPropsOnChange(
    ['path'],
    ({ update }) => {
      //update()
    }
  ),  
  withHandlers({
    mount: ({
      fromLayer,
      fromStage,
      instanceSet,
      update,
      x, y,
    }) => () => {
      const { stage } = fromStage

      const instance = new PIXI.Graphics()

      instance.position.set(x, y)
      instance.parentGroup = fromLayer.layer

      stage.addChild(instance)

      instanceSet(instance)

      update(instance)
    },
    unmount: ({
      fromStage,
      instance,
    }) => () => {
      const { stage } = fromStage
      if (stage && instance) {
        stage.removeChild(instance)
      }
    }    
  }),
  lifecycle({
    componentDidMount() {
      this.props.mount()
    },
    componentDidUpdate() {
      //this.props.update()
    },
    componentWillUnmount() {
      this.props.unmount()
    },
  }),
)

const Line2d = enhance(() => null)

export default Line2d
