// app/honey/page/web/PageProductList.js
import React from 'react'
import theme from 'app/honey/page/web/themeMUI';
import { Grid } from '@material-ui/core';
import Page from 'evoke-me/page/web/Page'
import SubHeader from 'app/honey/page/web/components/SubHeader'
import Poster from 'app/honey/page/web/components/Poster'
import Footer from 'app/honey/page/web/components/Footer'
import styled from 'styled-components'
import { StickyContainer, Sticky } from 'react-sticky';
import {
  CssBaseline,
  MuiThemeProvider,
  Typography,
  FormGroup,
  FormControlLabel,
  Switch,
  InputBase,
  Button,Tab,Tabs,

} from '@material-ui/core';

import {rem} from 'polished'

import {IconSearch} from './icons'
import ProductSampleList from './components/ProductSampleList'


const Title = styled(Typography)`
&& {
  color: ${props => props.textColor || '#fff'};
  margin: ${props => props.margin};
  font-weight: ${props => props.fweight || '700'};
  min-width: auto;
  width: 100%;
  font-size: ${rem(35)};
  text-align: center;
  font-size: ${rem(38)};
  text-transform: uppercase;

}
`;

const TabStyled = styled(Tab)`
&& {
  font-weight:normal;
  font-size: ${rem(14)};
  text-transform: uppercase;
  opacity: inherit;
  min-width: inherit;
  width: 50%;
  height: ${rem(48)};
  &:hover {
    background-color: #d8d8d8;
    color:black;
  }
  
  &:active {
    background-color: #d8d8d8;
    color:black;
    font-weight:bold;
  }
  
  &:first-child {
    border-right: 1px solid rgba(0,0,0,0.2);
  }
  
}
`;

const TabsStyled = styled(Tabs)`
&& {
  width:100%;
  margin-top:10px;
  margin-bottom:10px;
  background-color: #d8d8d8;
}
`;

const ReadMore = styled(({ size, align, margin, bold, color, ...other }) => (
  <Typography {...other} />
))`
&& {
  text-align: center;
  color: ${props => props.color || '#fff'};
  font-size: ${props => props.size};
  line-height: 1.4;
  font-weight: ${props => props.bold || 'normal'};
  font-style: italic;
  margin: ${props => props.margin || '0'};
  padding: ${props => props.padding || '0'};

  a {
    text-decoration: underline;
    color: ${props => props.color || '#fff'};
    font-weight: ${props => props.bold || 'normal'};
    &:hover {
      text-decoration: none;
    }
  }

}
`;

const TagLineContainer = styled.div`
&& {
  text-align: center;
  &.dark {
    padding: ${rem(70)} ${rem(45)};
    background-color: rgba(0, 0, 0, 0.2);
  }
}
`;

const TagLine = styled(({ color, transform, border, fweight, ...other }) => (
  <Typography {...other} />
))`
  && {
    font-size: ${rem(35)};
    font-weight: ${props => props.fweight || 400};
    color: ${props => props.color};
    text-align: center;
    font-style: normal;
    line-height: normal; 
    &.dark {
      padding: ${rem(70)} ${rem(45)};
      background-color: rgba(0, 0, 0, 0.2);

    }
  }
`;

const IconSearchStyled = styled(({ fontSize,margin, ...other }) => (
  <IconSearch {...other} />
))`
  && {
    font-size: inherit;
    margin: ${rem(7)};
    margin-top: ${rem(8)};
    
   
  }
`;

const InputBaseStyled = styled(({ fontSize,margin, ...other }) => (
  <InputBase {...other} />
))`
  && {
    font-size: inherit;
  }
`;


const FormGroupStyled = styled(({ fontSize, ...other }) => (
  <FormGroup {...other} />
))`
  && {
    font-size: ${rem(15)};
  }
`;

const ButtonStyled = styled(({ fontSize, ...other }) => (
  <Button {...other} />
))`
  && {
    font-size: ${rem(14)};
  }
`;
const SubHeaderWrapper = styled.div`
  && {
    z-index: 10;
    display: block;

  }
`; 

const FormControlLabelCustom = styled(FormControlLabel)`
  && {
    .label {
      font-size: ${rem(14)};
      display: flex;
    }
  }
`

const PageProductList = () => {
  return (
    <MuiThemeProvider theme={theme}>
      <CssBaseline />
      {/* CssBaseline is normalization for styles */}

      <Page hasFooter={true}>
        {/* Header */}

          <StickyContainer>   

            <Sticky>
              {({style}) => (
                <SubHeaderWrapper style={style} ><SubHeader /></SubHeaderWrapper>
                )}
            </Sticky>


            {/* Poster */}

            <Poster typical={'typical'} url={'/img/sample3.jpg'}>
              <Grid item xs={10} sm={8} alignItems={'center'} justify={'center'} >
                <TagLineContainer className={'dark'}>
                  <Title fweight="700" border="0" margin="0 0 1.25rem 0" transform="uppercase">
                    HONEY &amp; HIVE PRODUCTS
                  </Title>
                </TagLineContainer>
              </Grid>
            </Poster>


          {/* Filter */}


          <Grid container direction={'row'} justify="center" alignItems={'center'} spacing={0}>
              <Grid xs={10} item>
              
                <Grid
                  container
                  direction="row"
                  justify="center"
  alignItems="center" >
                  
                  <Grid xs={3} item>
                    <FormGroup row>
                      <FormControlLabelCustom control={<Switch value="checkedC" />} label={<span className={'label'}>Filter</span>} />
                    </FormGroup>
                  </Grid>

                  <Grid xs={4} item>
                  
                    <Grid 
                      container
                      direction="row"
                      justify="center"
                      alignItems="center"
                    >
                      <TabsStyled fullWidth>
 <TabStyled  fullWidth label="must have" />
                    <TabStyled fullWidth label="list all" />
                    </TabsStyled>
                      {/* <Grid item>
                        <ButtonStyled size={"large"} >must have</ButtonStyled>
                      </Grid>
                      | 
                      <Grid item>
                        <ButtonStyled size={"large"} >list all</ButtonStyled>
                      </Grid> */}

                    </Grid>

                    
                  </Grid>

                  <Grid xs={1} item> 
                  &nbsp;
                   </Grid>
                  <Grid xs={2} item> 
                  
                    <FormGroupStyled row>
                      <IconSearchStyled  />
                      <InputBaseStyled defaultValue="Search" />
                    </FormGroupStyled>

                   </Grid>

                </Grid>

              </Grid>
          </Grid>
      <br/>
      {/* line here */}
      <br/>
        {/* Product List */}

        <Grid container direction={'row'} justify="center" alignItems={'center'} spacing={0}>
              <Grid xs={10} item>

                  <Grid container direction={'row'} justify="center" alignItems={'center'} spacing={32}>
                    {(Array.apply(null, Array(12))).map((product, index) =>(
                    <Grid xs={4} item>
                      <ProductSampleList key={index} />
                    </Grid>
                    ))}
                  </Grid>


              </Grid>
          </Grid>


          </StickyContainer>
  
          {/* footer */}
          <Footer></Footer>


        </Page>
    </MuiThemeProvider>  
  )
}
export default PageProductList
