// app/honey/page/web/themeMUI.js
import { createMuiTheme } from '@material-ui/core';
import { orange, pink, green, grey, white } from "@material-ui/core/colors";

const theme = createMuiTheme({
    typography: {
      useNextVariants: true,
      fontFamily: [
        'Open Sans',
        'sans-serif',
      ].join(','),
      fontSize: 16,//dont change it
      htmlFontSize: 10,//font-size: 62.5%; /* 62.5% of 16px = 10px */
    },
    contrastThreshold: 3,
    tonalOffset: 0.2,
    palette: {
      primary: {
        main: '#080808',
        contrastText: '#fff'
      },
      secondary: {
        main: '#f5a623',
        contrastText: '#fff'
      },
      error: {
        main: '#f44336',
        contrastText: '#fff',   
      },
      action: {
        hover: '#fff',
        hoverOpacity: 0.08
      }
    },
    fontWeightMedium: 500,
    button: {
      borderRadius: 0,
    },
    overrides: {
      MuiButton: { // Name of the component ⚛️ / style sheet
        root: { // Name of the rule
          borderRadius: 0
        },
      },
      MuiTabs: {
        indicator: {
          backgroundColor: 'transparent'
        }
      },
      MuiTab: {
        root: {
          "&:hover": {
            backgroundColor: '#000000',
            color: '#ffffff'
          }
        },
        selected: {
          backgroundColor: '#000000',
          color: '#ffffff',
          "&:hover": {
            backgroundColor: '#000000',
            color: '#ffffff'
          }
        }
      }
    },

  });
  
  export default theme;
